servicetalk-http-router-jersey/src/main/java/io/servicetalk/http/router/jersey/DefaultJerseyStreamingHttpRouter.java

```diff
@@ -251,6 +259,120 @@ private void handle0(final HttpServiceContext serviceCtx, final StreamingHttpReq
+            return readEntity(rawType, annotations, getPropertiesDelegate());
+        }
+
+        public <T> T readEntity(final Class<T> rawType, final Type type) {
+            return readEntity(rawType, type, getPropertiesDelegate());
+        }
+
+        public <T> T readEntity(final Class<T> rawType, final Type type, final Annotation[] annotations) {
+            return readEntity(rawType, type, annotations, getPropertiesDelegate());
+        }
+
+        @Override
+        public <T> T readEntity(final Class<T> rawType, final Type type, final Annotation[] annotations,
+                                final PropertiesDelegate propertiesDelegate) {
+            final State prevState = state;
+            final boolean reentry = prevState == State.READING;
+            if (reentry || stateUpdater.compareAndSet(this, State.INIT, State.READING)) {
+                try {
+                    return super.readEntity(rawType, type, annotations, propertiesDelegate);
+                } finally {
+                    if (!reentry && !stateUpdater.compareAndSet(this, State.READING, State.INIT)) {
+                        // Closed while we were in progress.
+                        state = State.CLOSED;
+                        close0();
+                    }
+                }
+            }
+
+            throw new IllegalStateException(LocalizationMessages.ERROR_ENTITY_STREAM_CLOSED());
+        }
+
+        @Override
+        public boolean bufferEntity() throws ProcessingException {
+            final State prevState = state;
+            final boolean reentry = prevState == State.READING;
+            if (reentry || stateUpdater.compareAndSet(this, State.INIT, State.READING)) {
+                try {
+                    return super.bufferEntity();
+                } finally {
+                    if (!reentry && !stateUpdater.compareAndSet(this, State.READING, State.INIT)) {
+                        // Closed while we were in progress.
+                        state = State.CLOSED;
+                        close0();
+                    }
+                }
+            }
+
+            throw new IllegalStateException(LocalizationMessages.ERROR_ENTITY_STREAM_CLOSED());
+        }
+
+        @Override
+        public boolean hasEntity() {
+            if (state == State.CLOSED) {
+                throw new IllegalStateException(LocalizationMessages.ERROR_ENTITY_STREAM_CLOSED());
+            }
+
+            return super.hasEntity();
+        }
+
+        @Override
+        public InputStream getEntityStream() {
+            if (state == State.CLOSED) {
+                throw new IllegalStateException(LocalizationMessages.ERROR_ENTITY_STREAM_CLOSED());
+            }
+            return super.getEntityStream();
+        }
+
+        @Override
+        public void close() {
+            final State prevState = stateUpdater.getAndSet(this, State.PENDING_CLOSE);
+            if (prevState == State.INIT) {
+                state = State.CLOSED;
```

Comment:
consider moving `state = State.CLOSED;` into `close0` (it is common to all call sites).